mcproto/utils/abc.py

```diff
@@ -64,17 +75,58 @@ def __new__(cls: type[Self], *a, **kw) -> Self:
 
 
 class Serializable(ABC):
-    """Base class for any type that should be (de)serializable into/from :class:`~mcproto.Buffer` data."""
+    """Base class for any type that should be (de)serializable into/from :class:`~mcproto.Buffer` data.
+
+    Any class that inherits from this class and adds parameters should use the :func:`~mcproto.utils.abc.dataclass`
+    decorator.
+    """
 
     __slots__ = ()
 
-    @abstractmethod
+    def __post_init__(self) -> None:
+        """Run the validation method after the object is initialized."""
+        self.validate()
+        self.transform()
```

Comment:
hmmm, should transform really be called after validation? I'd think we would first want to transform here, so that validate can make some assumptions about the content being standardized.  I'm thinking of this method as essentially a replacement for `__init__` to do some extra logic, validation should only occur afterwards IMO.

(Note: if this gets changed, make sure to also update the docstring of both `transform` and `validate`, which mention the call order)

Reply:
I think both ways make sense here, but I decided on having `validate` first so that `transform` can assume the data is clean and doesn't have to throw an exception. 

I actually have in my code places where it would be nice is `transform` was first and places where the other is true. 

This is why I didn't separate them the first time.

In the light of this, what do you think I should do ?

Reply:
Okay, if in some cases, transform needs to be first, and in others it shouldn't be, it's a clear sign of overengineering here, drop `transform`, have subclasses override `__post_init__` instead, in which they can add their transform logic. It's then up to the subclass at which point to call `super().__post_init__`, which will internally only run validate.

This was one of the options in my previous suggestion on introducing the `transform` method, turns out it's probably gonna be the best solution.